pybind/mgr: Hack around the 'ImportError: PyO3 modules may only be in…

pecastro opened the pull request:
…itialized once per interpreter process' issue.

Not sure if this ( using /usr/bin/python3 ) is fair use given we can't really seem to be able to get the interpreter being used with 'sys.executable' under CPython.

Fixes: https://tracker.ceph.com/issues/64213

This is an alternative approach to the attempt #57926 at solving the problem.
User hashes don't change, and all the mgr_utils.py use cases have been dealt with.

It's not pretty, but it works.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

phlogistonjohn reviewed (commented):
This looks like a good start.
However, I strongly suggest taking the crypto code out of this file entirely and put it new .py file(s). Possibly in src/python-common somewhere and then instead of using `python3 -c`  use `python -m` with `if __name__ == '__main__` type guards in those files.

In the future we could then switch back to importing those new modules `if/when` the pyO3/rust/subintepreter issues are resolved with an even cleaner architecture overall.

Does this suggestion make sense to you?

pecastro commented:
It makes total sense however ...

When I had a go at trying to implement it I bumped into a side effect I had not seen before.
As soon as you move that code into a module that you run rather than run as a script,you are greeted with `DeprecationWarning: CSR support in pyOpenSSL is deprecated.` messages that will interfere with the existing logic to detect errors and the subsequent try/except order of things.

I've pushed the first stab for comparison and I provide a few examples of how to trigger things.

Run newly converted module and validate errors:
```
PYTHONPATH=/ceph/src/pybind/mgr:/ceph/src/pybind/mgr/.tox/py3/bin:/ceph/src/pybind:/usr/lib64/python312.zip:/usr/lib64/python3.12:/usr/lib64/python3.12/lib-dynload:/ceph/src/pybind/mgr/.tox/py3/lib64/python3.12/site-packages:/ceph/src/pybind/mgr/.tox/py3/lib/python3.12/site-packages:src/python-common \
python3 -m ceph.pybind.mgr.create_self_signed_cert eyJPIjogIkNlcGgiLCAiQ04iOiAibWdyIn0=

/ceph/build/src/python-common/ceph/pybind/mgr/create_self_signed_cert.py:18: DeprecationWarning: CSR support in pyOpenSSL is deprecated. You should use the APIs in cryptography.
  req = crypto.X509Req()
b'-----BEGIN CERTIFICATE-----
....

```

pyOpenSSL versions
```
(mgr-virtualenv) [pecastro@stealthix build]$ PYTHONPATH=/ceph/src/pybind/mgr:/ceph/src/pybind/mgr/.tox/py3/bin:/ceph/src/pybind/mgr/$PYTHONPATH:/ceph/src/pybind:/usr/lib64/python312.zip:/usr/lib64/python3.12:/usr/lib64/python3.12/lib-dynload:/ceph/src/pybind/mgr/.tox/py3/lib64/python3.12/site-packages:/ceph/src/pybind/mgr/.tox/py3/lib/python3.12/site-packages:src/python-common  python -m OpenSSL.debug
pyOpenSSL: 25.0.0
cryptography: 44.0.0
cffi: 1.17.1
cryptography's compiled against OpenSSL: OpenSSL 3.4.0 22 Oct 2024
cryptography's linked OpenSSL: OpenSSL 3.4.0 22 Oct 2024
Python's OpenSSL: OpenSSL 3.2.2 4 Jun 2024
Python executable: /ceph/build/mgr-virtualenv/bin/python
Python version: 3.12.7 (main, Oct  1 2024, 00:00:00) [GCC 14.2.1 20240912 (Red Hat 14.2.1-3)]
Platform: linux
sys.path: ['/ceph/build', '/ceph/src/pybind/mgr', '/ceph/src/pybind/mgr/.tox/py3/bin', '/ceph/src/pybind', '/usr/lib64/python312.zip', '/usr/lib64/python3.12', '/usr/lib64/python3.12/lib-dynload', '/ceph/src/pybind/mgr/.tox/py3/lib64/python3.12/site-packages', '/ceph/src/pybind/mgr/.tox/py3/lib/python3.12/site-packages', '/ceph/build/src/python-common', '/ceph/build/mgr-virtualenv/lib64/python3.12/site-packages', '/ceph/build/mgr-virtualenv/lib/python3.12/site-packages']
```

So, the options I think we have are:

1. Leave it as it is. Document the reasons, Open story ticket to deal with it properly or wait for a fix ....
2. Continue the refactoring into a separate module but change the logic to accommodate the error in some way shape or form.
3. Substitute all the deprecated pyOpenSSL code by the pyca/cryptography module, there is prior art in ./src/pybind/mgr/cephadm/ssl_cert_utils.py that at a very quick glance could be used as reference.
4. Something else I'm not thinking of ?

I'm leaning towards option 3.

pecastro commented:
:man_facepalming:  I never pushed the first stab ... 

I reworked everything following your initial suggestion and applied the warnings, it turned out only the first block of code had the need for it. 
We can leave the substitution for the pyca/cryptography module for a later stage.

pecastro commented:
WOOM.

```
[pecastro@host build]$ rm -rf dev/ out/                                                           
[pecastro@host build]$ MDS=1 MGR=1 MON=1 OSD=1 NFS=0 ../src/vstart.sh -n                                                                                
rm -f core*                                                                                                                                                                                                  
hostname host                                                                                                                                                                                           
ip 192.168.12.74                                                                                                                                                                                             
port 40915                                                                                                                                                                                                   
/ceph/build/bin/ceph-authtool --create-keyring --gen-key --name=mon. /ceph/build/keyring --cap mon 'allow *'                                                                
creating /ceph/build/keyring                                                                                                                                                                                 
/ceph/build/bin/ceph-authtool --gen-key --name=client.admin --cap mon 'allow *' --cap osd 'allow *' --cap mds 'allow *' --cap mgr 'allow *' /ceph/build/keyring             
/ceph/build/bin/monmaptool --create --clobber --addv a [v2:192.168.12.74:40915,v1:192.168.12.74:40916] --print /tmp/ceph_monmap.1686593                                     
/ceph/build/bin/monmaptool: monmap file /tmp/ceph_monmap.1686593                                                                                                                                             
setting min_mon_release = quincy                                                                                                                                                                             
/ceph/build/bin/monmaptool: set fsid to b4739f56-ef23-4af4-b7b2-d7a41303e75c                                                                                                                                 
epoch 0                                                                                                                                                                                                      
fsid b4739f56-ef23-4af4-b7b2-d7a41303e75c                                                                                                                                                                    
last_changed 2025-02-13T23:36:50.171548+0000                                                                                                                                                                 
created 2025-02-13T23:36:50.171548+0000                                                                                                                                                                      
min_mon_release 17 (quincy)
...
...
vstart cluster complete. Use stop.sh to stop. See out/* (e.g. 'tail -f out/????') for debug output.

dashboard urls: https://192.168.12.74:41915
  w/ user/pass: admin / admin

export PYTHONPATH=/ceph/src/pybind:/ceph/build/lib/cython_modules/lib.3:/ceph/src/python-common:$PYTHONPATH
export LD_LIBRARY_PATH=/ceph/build/lib:$LD_LIBRARY_PATH
export PATH=/ceph/build/bin:$PATH
export CEPH_CONF=/ceph/build/ceph.conf
alias cephfs-shell=/ceph/src/tools/cephfs/shell/cephfs-shell
CEPH_DEV=1
[pecastro@host build]$ ceph status
2025-02-13T23:38:03.196+0000 7f81085456c0 -1 WARNING: all dangerous and experimental features are enabled.
2025-02-13T23:38:03.203+0000 7f81085456c0 -1 WARNING: all dangerous and experimental features are enabled.
  cluster:
    id:     b4739f56-ef23-4af4-b7b2-d7a41303e75c
    health: HEALTH_WARN
            4 pool(s) have no replicas configured
 
  services:
    mon: 1 daemons, quorum a (age 72s)
```
![image](https://github.com/user-attachments/assets/dcdf9932-8293-4b7d-b70e-e7b436fd0fa5)

pecastro commented:
jenkins test api

pecastro commented:
jenkins test make check

pecastro commented:
jenkins test dashboard cephadm

pecastro commented:
jenkins test dashboard cephadm

pecastro commented:
@nizamial09 @epuertat Given you both did some work in this area maybe you can advise forward or to /dev/null.

rkachach commented:
@pecastro @phlogistonjohn related to the crypto functions (such as `create_self_signed_cert`) as @pecastro found out we already have support in `cephadm/ssl_cert_utils.py` to generate self-signed certificates. Besides, the above function is only used by the dashboard in the following code:

```
    @CLIWriteCommand("dashboard create-self-signed-cert")
    def set_mgr_created_self_signed_cert(self):
        cert, pkey = create_self_signed_cert('IT', 'ceph-dashboard')
        result = HandleCommandResult(*self.set_ssl_certificate(inbuf=cert))
        if result.retval != 0:
            return result

        result = HandleCommandResult(*self.set_ssl_certificate_key(inbuf=pkey))
        if result.retval != 0:
            return result
        return 0, 'Self-signed certificate created', ''
```
Which doesn't make sense to me as cephadm now is the responsible for generating self-signed certificates for the any mgr module who needs that (cephadm provides `orch certmgr generate-certificates` which is already used by `dashboard` and `prometheus mgr-module`) ... so moving forward I think we should be able to move away from using `OpenSSL` and switch to `cryptography` library probably making the code in `cephadm/ssl_cert_utils.py` accessible in python-common so other modules can use it.

phlogistonjohn commented:
> @pecastro @phlogistonjohn related to the crypto functions (such as `create_self_signed_cert`) as @pecastro found out we already have support in `cephadm/ssl_cert_utils.py` to generate self-signed certificates. Besides, the above function is only used by the dashboard in the following code:
> 
> ```
>     @CLIWriteCommand("dashboard create-self-signed-cert")
>     def set_mgr_created_self_signed_cert(self):
>         cert, pkey = create_self_signed_cert('IT', 'ceph-dashboard')
>         result = HandleCommandResult(*self.set_ssl_certificate(inbuf=cert))
>         if result.retval != 0:
>             return result
> 
>         result = HandleCommandResult(*self.set_ssl_certificate_key(inbuf=pkey))
>         if result.retval != 0:
>             return result
>         return 0, 'Self-signed certificate created', ''
> ```
> 
> Which doesn't make sense to me as cephadm now is the responsible for generating self-signed certificates for the any mgr module who needs that (cephadm provides `orch certmgr generate-certificates` which is already used by `dashboard` and `prometheus mgr-module`) ... so moving forward I think we should be able to move away from using `OpenSSL` and switch to `cryptography` library probably making the code in `cephadm/ssl_cert_utils.py` accessible in python-common so other modules can use it.

I'm fine with those suggestions but I also think we need to take things one step at a time. On top of that, `cryptography` is a library that brings in rust dependencies and would be subject to this interpreter problem. So I think this effort is largely justified still. I don't know how much we want to put on @pecastro who is largely just trying to fix the pyO3 side of things AFAIK.

pecastro commented:
What  @phlogistonjohn said.
> So I think this effort is largely justified still. I don't know how much we want to put on @pecastro who is largely just trying to fix the pyO3 side of things AFAIK.

But mostly this:

> On top of that, `cryptography` is a library that brings in rust dependencies and would be subject to this subinterpreter problem.

This MR offers a potential solution for what is an existing problem and it does so in a way that is modular and feasible of rework in the future.
The next step could be refactoring these and substituting them by cryptography functions provided that by then they already work in  a subinterpreter or by applying a similar adapter pattern to has been done here.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -627,63 +627,48 @@ def create_self_signed_cert(organisation: str = 'Ceph',
 
-    return cert.decode('utf-8'), pkey.decode('utf-8')
+    cert, pkey = result.stdout.strip().split("#######")
+    return eval(cert).decode('utf-8'), eval(pkey).decode('utf-8')
```
`eval()` is the root of all evils and should eradicated from all programming languages and buried 100 kms beneath the Earth's crust... and mostly 'cause I don't understand why it's needed here.

If someone manages to inject a file matching the `ceph.pybind.mgr...` spec with higher prio in the PYTHONPATH then they'll have `ceph-mgr` executing random code as the `ceph` user. To give you a hint, this is `PYTHONPATH` inside ceph-mgr (and this is platform-dependent):
```
./src/pybind/mgr
/usr/local/lib64/python3.10/site-packages
/usr/local/lib/python3.10/site-packages
/usr/lib64/python3.10/site-packages
/usr/lib/python3.10/site-packages
./src/pybind
./build/lib/cython_modules/lib.3
./src/python-common

/usr/lib64/python310.zip
/usr/lib64/python3.10
/usr/lib64/python3.10/lib-dynload
```

nh2 replied:
As a user and commenter on https://tracker.ceph.com/issues/64213, I just want to support this feedback:

It is greatly appreciated that the PyO3 problem is being fixed :+1:, but please without `eval()`, execution-from-string, or secrets visible in `ps`.

Thank you all who work on this!

pecastro replied:
`eval` can be useful, but it is indeed EVIL!
This new revision I've just pushed, though a bit rushed, hopefully takes care of all 3 issues.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -10,7 +10,7 @@
     import tests  # noqa
 
-import bcrypt
+import base64
```
Same thing, can we avoid base64-encoding stuff by passing binary data instead? We shouldn't pass sensitive (not even secrets, just hints) through the CLI.

pecastro replied:
Done.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -627,63 +627,48 @@ def create_self_signed_cert(organisation: str = 'Ceph',
+        raise ValueError(result.stderr)
 
-    return cert.decode('utf-8'), pkey.decode('utf-8')
+    cert, pkey = result.stdout.strip().split("#######")
```
Why not creating 2 different calls, one for generating the Pkey, and another for the cert (passing the key as input arg)?

pecastro replied:
Done.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -707,50 +692,40 @@ def verify_cacrt(cert_fname):
+        if result.returncode != 0 or result.stderr != '':
+            raise ValueError(result.stderr)
+
+        org_name, cn = result.stdout.strip().split("#######")
```
Same thing here, 2 different calls, one for getting the org name, and another for the canonical name.

pecastro replied:
Done

epuertat commented on src/python-common/ceph/pybind/mgr/create_self_signed_cert.py:
```diff
@@ -0,0 +1,49 @@
+from uuid import uuid4
+
+
+def main() -> None:
+    dname = json.loads(base64.b64decode(sys.argv[1]).decode("utf-8"))
+
+    # create a key pair
+    pkey = crypto.PKey()
+    pkey.generate_key(crypto.TYPE_RSA, 2048)
```
Instead of having 4 new files (modules), why not putting everything inside a single module and using argparse to select the proper method?:
```sh
$ python3 -m ceph.pybind.mgr.cryptotools create_self_signed_cert --private-key
# output: private key

$ echo private_key | python3 -m ceph.pybind.mgr.cryptotools create_self_signed_cert --certificate 
# output: certificate

...
```

This is slightly more complex, but saves the hassle of new 4 files...

pecastro replied:
Done.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -978,8 +943,27 @@ def wrapper(*args: Any, **kwargs: Any) -> T:
-    return bcrypt.hashpw(password.encode('utf8'), salt).decode('utf8')
+        salt_password = ''
+
+    __bcrypt_password = """import bcrypt
+
+password = "{}"
+salt_password = "{}"
+
+if not salt_password:
+    salt = bcrypt.gensalt()
+else:
+    salt = salt_password.encode('utf8')
+
+print(bcrypt.hashpw(password.encode('utf8'), salt).decode('utf8'))""".format(password, salt_password)
+
+    import subprocess
+    result = subprocess.run(["/usr/bin/python3", "-c", __bcrypt_password], capture_output=True, text=True)
```
Oh, I think I missed this part... This behaves the same as an `eval()/exec()` call, and is worse because if fed with user supplied data, so this is basically introducing a code-injection vulnerability.

What happens, for example, if a user tries to log in with  the password `";import subprocess;subprocess.run("sudo rm -rf /var/lib/ceph");`?

pecastro replied:
That bit of code was "cough" "cough" ... highly experimental. 
It was part of the initial framework I put together just to prove that it could be done. I'm very glad you spotted it.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -627,63 +627,48 @@ def create_self_signed_cert(organisation: str = 'Ceph',
-            raise ServerConfigException(msg)
-    except (ValueError, crypto.Error) as e:
+        import subprocess
+        result = subprocess.run(["python3", "-m", "ceph.pybind.mgr.verify_cacrt_content",
```
BTW, we should see if we can guess the python interpreter from inside the Python environment. Something like:
```py
import sys
 
print(sys.executable)
# /usr/bin/python
```
Not sure what that will print inside a ceph-mgr Python sub-interpreter though...

pecastro replied:
That will print empty. I started with that.

pecastro commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -627,63 +627,48 @@ def create_self_signed_cert(organisation: str = 'Ceph',
+        import subprocess
+        result = subprocess.run(["python3", "-m", "ceph.pybind.mgr.verify_cacrt_content",
+                                 base64.b64encode(crt if isinstance(crt, bytes) else crt.encode('ascii')).decode("ascii")],
+                                capture_output=True, text=True)
```
We don't and we can make it work with `binary`.

pecastro commented:
Guys, its still overall not pretty, but it looks a lot better than before.

This new `cryptotools.py` module needs to die soon, not only the hacky approach is smelly, there's lots of other hidden gems like the warnings across a bunch of methods that will soon be deprecated, and whatever else I might have introduced with the rewrite from 4 modules into 1.

If you decide to merge this let's please get a tech debt ticket in so it doesn't get forgotten.

pecastro commented:
jenkins test make check

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -627,70 +626,59 @@ def create_self_signed_cert(organisation: str = 'Ceph',
-    # populate the subject with the dname settings
-    for k, v in dname.items():
-        setattr(subj, k, v)
+    import json
+    import subprocess
```
Let's import these at the bottom of the file.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -627,70 +626,59 @@ def create_self_signed_cert(organisation: str = 'Ceph',
+    import json
+    import subprocess
+
+    private_key = subprocess.run(["python3", "-m", "ceph.pybind.mgr.cryptotools", "create_self_signed_cert", "--private_key"],
```
In order to facilitate unit testing and mocking of the multiple `subprocess.run()` method, we can encapsulate those in wrapper methods:
```suggestion
    private_key = create_private_key()
    ...
    cert = create_self_signed_cert(private_key=private_key, o=organisation, cn=common_name)
```

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -627,70 +626,59 @@ def create_self_signed_cert(organisation: str = 'Ceph',
-
-    except (ValueError, crypto.Error) as e:
+        import subprocess
+        result = subprocess.run(["python3", "-m", "ceph.pybind.mgr.cryptotools", "verify_cacrt_content"],
+                                input=crt if isinstance(crt, bytes) else crt.encode('utf-8'),
+                                capture_output=True)
+        # The above script will only produce stdout output.
+        # The only scenarios that produce stderr output are failures to import modules
+        # or syntax errors which test_tls.py will catch
+
+        # Check result of CompletedProcess
+        if result.returncode != 0 or result.stderr != b'':
+            logger.warning(result.stderr)
+            raise ValueError(result.stderr)
```
Given this structure (`run ; if .returncode ...`) repeats a lot, I'd wrap it inside a helper method.

epuertat commented on src/python-common/ceph/pybind/mgr/cryptotools.py:
```diff
@@ -0,0 +1,197 @@
+        print(f'Invalid cert/key pair: {e}')
+
+
+if __name__ == "__main__":
+    # create the top-level parser
+    parser = argparse.ArgumentParser(prog='cryptotools.py')
+    subparsers = parser.add_subparsers(required=True)
+
+    # create the parser for the "password_hash" command
+    parser_foo = subparsers.add_parser('password_hash')
+    parser_foo.set_defaults(func=password_hash)
+
+    # create the parser for the "create_self_signed_cert" command
+    parser_bar = subparsers.add_parser('create_self_signed_cert')
+    parser_bar.add_argument('--private_key', required=False, action='store_true')
+    parser_bar.add_argument('--certificate', required=False, action='store_true')
+    parser_bar.set_defaults(func=create_self_signed_cert)
+
+    # create the parser for the "verify_cacrt_content" command
+    parser_bar = subparsers.add_parser('verify_cacrt_content')
+    parser_bar.set_defaults(func=verify_cacrt_content)
+
+    # create the parser for the "get_cert_issuer_info" command
+    parser_bar = subparsers.add_parser('get_cert_issuer_info')
+    parser_bar.add_argument('--org_name', required=False, action='store_true')
+    parser_bar.add_argument('--cn', required=False, action='store_true')
+    parser_bar.set_defaults(func=get_cert_issuer_info)
+
+    # create the parser for the "verify_tls" command
+    parser_bar = subparsers.add_parser('verify_tls')
+    parser_bar.set_defaults(func=verify_tls)
+
+    # parse the args and call whatever function was selected
+    args = parser.parse_args()
+    args.func(args)
```
Awesome clean-up, @pecastro ! 🥂

pecastro replied:
Tkx.

epuertat reviewed (commented):
Awesome work in such a short time! Thanks a lot, @pecastro !

Some comments:
* As you mention that this should be just a short/mid-term work-around, perhaps we should keep the original code in place and switch the new implementation via some `if PyO3_issue == True:` toggles.
* I've been looking at how to simplify the string-binary encoding decoding... Probably having some wrapping code that does all the encoding/decoding before/after calling `subprocess.run`.
* Also, default encoding/decoding in Python is `utf-8`, so probably we can leave it non-explicit (`str.encode()`, `bytes.decode()`).

I can't approve this yet (not my best day today to ensure a decent review), so I'll defer/wait for others to review this as well, and will revisit throughout this week.

nizamial09 commented:
jenkins test dashboard

nizamial09 commented:
jenkins test dashboard

nizamial09 commented:
jenkins test dashboard cephadm

nizamial09 commented:
jenkins test dashboard

phlogistonjohn commented:
Hi all, this is really shaping up well. There are a few a few of the items @epuertat noted such as to encapsulate the calls to subprocess.run in wrapper methods and consistently checking the result to these subprocesses that ought to be implemented.  There are a few other small things like the location of the new file and some variable names I will suggest updates to. However, I heard that @pecastro is away for some time. 

I'm willing to provide some patches directly on top of this effort - OR - am happy to just provide some feedback as comments. Let me know what folks are thinking here. In the meantime @adk3798 and I think we have a reproducer for the pyo3 issue and I'll be testing that out and testing this patch on top of that over the next couple of days.

I'll report here what I find.

epuertat commented:
Thanks a lot, @phlogistonjohn! Please, feel free to contribute, since we don't know when @pecastro is back and this is becoming an urgent matter with the recent upgrades of `python3-cryptography` package due to CVEs.

Regarding the reproducer, let me summon @nizamial09 here, since he was setting up a Fedora 41 environment for that purpose.

nizamial09 commented:
> Regarding the reproducer, let me summon @nizamial09 here, since he was setting up a Fedora 41 environment for that purpose.

I have a Dockerfile to build a ceph cluster on top of [fedora41 here](https://github.com/nizamial09/ceph-rpm-fedora). And I built one and if you simply start a container like `docker run -it nizamial09/ceph-rpm:fedora` you can reproduce the issue.

The problem with my setup is that it runs ceph 19.2.1 since that's the one available in official fedora. And when I tried to run 20.0.x it failed because of missing dependencies and I couldn't fix it for now. So I just left it as it is. To test this PR I cherry-picked it on top of a squid branch and started the container like

```
docker run -it \                                 
  -v /home/nia/projects/ceph/src/pybind/mgr:/usr/share/ceph/mgr \
  -v /home/nia/projects/ceph/src/python-common:/usr/share/ceph/python-common \
  -e PYTHONPATH=/usr/share/ceph/python-common \
  nizamial09/ceph-rpm:fedora
```

which mounted the `mgr` and `python-common` to the container and I see dashboard module loaded ok but saw this in the logs for cephadm

```
2025-04-15T13:28:43.829+0000 7ff1885c9480  1 mgr[py] Loading python module 'cephadm'
2025-04-15T13:28:44.058+0000 7ff1885c9480 -1 mgr[py] Module not found: 'cephadm'
2025-04-15T13:28:44.058+0000 7ff1885c9480 -1 mgr[py] Traceback (most recent call last):
  File "/usr/share/ceph/mgr/cephadm/__init__.py", line 1, in <module>
    from .module import CephadmOrchestrator
  File "/usr/share/ceph/mgr/cephadm/module.py", line 40, in <module>
    from cephadm.http_server import CephadmHttpServer
  File "/usr/share/ceph/mgr/cephadm/http_server.py", line 6, in <module>
    from cephadm.agent import AgentEndpoint
  File "/usr/share/ceph/mgr/cephadm/agent.py", line 23, in <module>
    from cephadm.ssl_cert_utils import SSLCerts
  File "/usr/share/ceph/mgr/cephadm/ssl_cert_utils.py", line 8, in <module>
    from cryptography import x509
  File "/usr/lib64/python3.13/site-packages/cryptography/x509/__init__.py", line 7, in <module>
    from cryptography.x509 import certificate_transparency, verification
  File "/usr/lib64/python3.13/site-packages/cryptography/x509/certificate_transparency.py", line 11, in <module>
    from cryptography.hazmat.bindings._rust import x509 as rust_x509
ImportError: PyO3 modules compiled for CPython 3.8 or older may only be initialized once per interpreter process

2025-04-15T13:28:44.059+0000 7ff1885c9480 -1 mgr[py] Class not found in module 'cephadm'
2025-04-15T13:28:44.059+0000 7ff1885c9480 -1 mgr[py] Error loading module 'cephadm': (2) No such file or directory
2025-04-15T13:28:44.059+0000 7ff1885c9480  1 mgr[py] Loading python module 'cli_api'
2025-04-15T13:28:44.114+0000 7ff1885c9480 -1 mgr[py] Module cli_api has missing NOTIFY_TYPES member
2025-04-15T13:28:44.114+0000 7ff1885c9480 -1 mgr[py] Module not found: 'mgr_module'
2025-04-15T13:28:44.114+0000 7ff1885c9480 -1 mgr[py] AttributeError: type object 'CLI' has no attribute 'NOTIFY_TYPES'

2025-04-15T13:28:44.114+0000 7ff1885c9480  1 mgr[py] Loading python module 'crash'
2025-04-15T13:28:44.169+0000 7ff1885c9480 -1 mgr[py] Module crash has missing NOTIFY_TYPES member
2025-04-15T13:28:44.169+0000 7ff1885c9480 -1 mgr[py] Module not found: 'mgr_module'
2025-04-15T13:28:44.169+0000 7ff1885c9480 -1 mgr[py] AttributeError: type object 'Module' has no attribute 'NOTIFY_TYPES'

2025-04-15T13:28:44.170+0000 7ff1885c9480  1 mgr[py] Loading python module 'dashboard'

```

I don't know whether its because I am doing it from squid.

Here's the docker tag incase its useful `docker pull nizamial09/ceph-rpm:fedora`

phlogistonjohn commented:
https://github.com/ceph/ceph/compare/main...phlogistonjohn:ceph:jjm-pe-crypto-plus

I've added a series of patches on top of @pecastro's current patch in this PR. The work can be viewed at the branch link above. The end result is the same in spirit - I just tried to create an intermediate abstraction between the functions-as-subprocesses and the mgr code. I cleaned up a few interfaces, tried to use JSON whenever non-blob values are exchanged and slightly reorganized the locations of some of the files.

Let me know how we ought to proceed with this - perhaps I can file a new PR or is there other way we can collaborate? Regardless, my next step is to test - some basic local testing has looked good but we should also be prepared to test in teuthology. In particular we should do runs with and without cryptography packages that trigger the pyo3 import error.

epuertat commented:
@phlogistonjohn had a quick look and didn't see any immediate issue. Is this still WIP? Asking 'cause I still think that we should keep the old implementation switchable with a `PYO3_ISSUE` feature toggle. Although this is not performance critical code, having to instantiate a new interpreter for each crypto op is not ideal either.

pecastro commented:
Hi all.
I went MIA on account of family holidays booked.

I'm slowly catching up on some of the nice work @phlogistonjohn did which handles mostly all the of the previous cases you @epuertat had spotted as in need of improvement.  

I've just fixed a small typo and pushed and I suspect most of the remaining failures are due to the expectation of bytes vs string across the remote.py plumbing.

The extra shim "has a different opinion" about the type of data returned that might have deeper implications. I like the standardization of returning strings for strings sake but some of the caller functions/methods might not agree with it.
e.g. https://github.com/ceph/ceph/blob/f2edb7268da39edf214c920a84dd4c65c628f272/src/pybind/mgr/dashboard/module.py#L417C44-L417C67 

I'm running tests locally to see if I understand the full implications of this as well as the bytes vs strings mismatches.

pecastro commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -627,70 +626,59 @@ def create_self_signed_cert(organisation: str = 'Ceph',
-
-    except (ValueError, crypto.Error) as e:
+        import subprocess
+        result = subprocess.run(["python3", "-m", "ceph.pybind.mgr.cryptotools", "verify_cacrt_content"],
```
That, ( returning days)  was a relatively recent addition to the method that also caught me off guard when I re-did the work, it used to be a true or false return.

https://github.com/ceph/ceph/commit/3bb6c572a380c94f3a6b8e57f0bc5c5cde5e3c79#diff-ba538532bb5450d415ab80916916bd68fe51d195e57c9aa34effaa8789d7a392

I'll look into it.

pecastro replied:
Done.

epuertat commented:
@phlogistonjohn I was thinking on either:
* a centralized mgr config option (env. vars are not that nice at cluster scale), or
* a runtime check:
  * `try: import cryptography; except ImportError: PYO3_ISSUE= True;`, but that would be problematic since it'd follow 2 different approaches on 1st and 2nd+ runs :/
  * `if Version("41.0.0") <= Version(cryptography.__version__) <= Version("Version without Pyo3 sub-interpreter issue")...`, but we don't know yet when the `pyo3` project will handle this issue...

So my vote would be for the centralized config option.

phlogistonjohn commented:
OK, that might be tricky wrt to the dependency flow, but I'll look into it. Right now we're just testing the basic workflow and yesterday we found another use of bcrypt module in the dashboard that was not being handled by the external process. I'm waiting to hear back on the status of that test run to see if we have more imports that need to be cleaned up. If all goes well, I'll continue organizing/refactoring.

phlogistonjohn commented:
A brief update: @adk3798 ran the combined patch series (plus a fixup commit) through teuthology and it failed again. After almost a whole day's investigation we determined that the kubernetes module, used by k8sevents and rook mgr modules, imports cryptography and if either of these were imported before cephadm the boostrap would fail.  We've summarized some proposed solutions here: https://pad.ceph.com/p/pyo3-cryptography-imports-k8s
For now I will continue helping with this effort to isolate the dashboard from direct uses of cryptography but this is also something that needs considieration. Because one of the options is to remove the k8sevents and rook mgr modules we plain on raising it as a topic at the next CSC meeting (Apr. 28th).

pecastro commented:
@epuertat  When I said the cryptotools needed to die I was talking from a general perspective as in with hope that the issue would be solved by the pyo3 subintepreters folks.
As @phlogistonjohn  had suggested earlier in the conversation the process of splitting things out into a module would then make that job easier.

I understand the logic behind keeping the original code but have we considered the cons of this "suggestion"? 
There will be two code paths to maintain/manage for as long as the pyo3 bug rages on, this is extra complexity being added.
These code path doesn't seem to be in need to be that high performant, I mean how many certs a minute are we talking of issuing and validating ?
Some food for thought.

I've added and a few more commits, some of which might be a bit controversial, like the last one removing ascii and utf-8 references across encode/decode. The tests seem to pass without those being explicitly stated.
Regarding my earlier comment, I hadn't noticed that everything is already being passed as strings with the exception of `verify_tls` function in mgr_util which has a test that specifies a bytes based primary key.
So, as long as the return values matches expectations which they seem to based on the testing it should be good.

nizamial09 commented:
> one of the options is to remove the k8sevents

i had a tracker months ago to deprecate it since no one is actively maintaing it: https://tracker.ceph.com/issues/66091

phlogistonjohn commented:
Hi all, I have updated my branch with additional fixes and rebased on main. Unfortunately, one of the things I feared is starting to happen - @pecastro and I are fixing the same set of issues but in different ways. For example - I updated a test to stop encoding a string to bytes while @pecastro changed the code to accept either strings or bytes.  Obviously, I prefer my approach and I think I can justify it but we're doing redundant work and creating extra work to reconcile the changes.

To make my changes more visible I am going to create a seperate (draft) PR. My plan is to use it to partially rework the interface to generalize it and allow for either "subprocess mode" or "internal mode" - as discussed with @epuertat earlier in this PR. We will have to come up with a plan to reconcile our changes at some point though.

phlogistonjohn commented:
My branch, in PR form: #62951

pecastro commented:
> For example - I updated a test to stop encoding a string to bytes while @pecastro changed the code to accept either strings or bytes. Obviously, I prefer my approach and I think I can justify it but we're doing redundant work and creating extra work to reconcile the changes.

@phlogistonjohn curiously my first fix was the same as yours :smile:  , in the meantime before pushing, I added the encoding to honour the test thinking , who knows, maybe there's a legitimate case somewhere where a bytes pkey is passed in.
I'm not married to it.
I know the tests will all pass either way.

Given you're also in flow I can pause / stand down to minimize noise / attrition leaving the remaining bit of changing that `verify_cacrt_content` function to last. 

In case you feel we should talk, I exist on a GMT schedule, but most of the work I do here is on GMT evenings.

pecastro commented:
jenkins test api

phlogistonjohn commented:
> > For example - I updated a test to stop encoding a string to bytes while @pecastro changed the code to accept either strings or bytes. Obviously, I prefer my approach and I think I can justify it but we're doing redundant work and creating extra work to reconcile the changes.
> 
> @phlogistonjohn curiously my first fix was the same as yours 😄 , in the meantime before pushing, I added the encoding to honour the test thinking , who knows, maybe there's a legitimate case somewhere where a bytes pkey is passed in. I'm not married to it. I know the tests will all pass either way.

:-)  Agreed. The reason I prefer to change the test and not the library code is that mypy is run (only) on the modules dir and passes indicating that there is a very good chance no other production code is passing in a bytes array. Since tests exist to verify the behavior of the production code and 

> 
> Given you're also in flow I can pause / stand down to minimize noise / attrition leaving the remaining bit of changing that `verify_cacrt_content` function to last.
> 

I appreciate your offer, but I don't think I need that (yet). I just won't be rebasing on your branch frequently for the next couple of work days - not until all the fetures @epuertat and I discussed are implemented. I end up spending too much time in rebase, resolving conflicts rather than updating the code. I think it's fine if you continue to test and prove out that your patches work for your use cases. I'm working on making the code more modular and then making it configurable. In the meantime @adk3798 is helping test things in teuthology - main to prove the concept works as well as find other things we had previously overlooked, like the rook/k8sevents issue.

Perhaps late next week I will revisit what you have done in the meantime and reconcile the branches. 


> In case you feel we should talk, I exist on a GMT schedule, but most of the work I do here is on GMT evenings.

Thank you for the offer. I don't think I need that (yet). More generally, if you are interested there's a [Orch Weekly every Tuesday](https://pad.ceph.com/p/orchestration-weekly) where we discuss design and various related topics and you are certainly welcome to join one of those too.

pecastro commented:
> I think it's fine if you continue to test and prove out that your patches work for your use cases.

I'm done, obviously pending a re-review. 

The tests are passing and other than the dual mode running feature which @phlogistonjohn seems to be working on I've proven that this workaround is feasible to implement which was my primary goal with this MR alongside the perceived urgency of this fix.

I've added the function name renaming that @epuertat  suggested, and amended the commit messages to the standard but everything should still pretty much be as when @phlogistonjohn last rebased.
